### DISCUSSION TITLE
When should we create an instance of PrismaClient?

### DISCUSSION THREAD

**Discussion by alan345** [prisma/prisma#3209] [Q&A] [answered]
lets assume we have 2 functions. What is the good practice? 

```ts
test: async () => {
    const prisma = new PrismaClient() // New instance
    const id = await getUserId()
    return prisma.test.findOne({where: { id } })
}

getUser: async () => {
    const prisma = new PrismaClient() // New instance
    const user = await prisma.user.findOne({ where: { id: 'userId' } })
    return user
}
```

or

```ts
test: async () => {
    const prisma = new PrismaClient()
    const id = await getUserId(prisma)
    return prisma.test.findOne({where: { id } })
}

getUser: async (prisma) => {
    const user = await prisma.user.findOne({ where: { id: 'userId' } })
    return user
}
```

Should we create a new instance of PrismaClient in each function or pass `prisma` as an argument?
what will be the type for `prisma`? 

```ts
getUser: async (prisma: WhatIsTheType) => {
    const user = await prisma.user.findOne({ where: { id: 'userId' } })
    return user
}
```

**Comment by ryands17** [comment, marked as answer]
Hey @alan345 👋 

Prisma Client should only be initialized once, outside any functions so the entire app uses only a single instance as follows:

```ts
const prisma = new PrismaClient()

test: async () => {
    const id = await getUserId(prisma)
    return prisma.test.findOne({where: { id } })
}

getUser: async () => {
    const user = await prisma.user.findOne({ where: { id: 'userId' } })
    return user
}
```

And if you still require to pass the Prisma Client as an argument, you can use:

```ts
getUser: async (prisma: PrismaClient) => {
    const user = await prisma.user.findOne({ where: { id: 'userId' } })
    return user
}
```

**Reply by alan345** [reply]
Ok thank! 

With a graphQL server, without Nexus, should I pass the Prisma instance in the context?

And use it in my resolvers like: `ctx.prisma.user.findOne(..)`?

**Reply by ryands17** [reply]
Yes that would be the best approach :)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
